### ISSUE TITLE
Discriminator with oneOf results in discriminator properties nested in child properties.

### ISSUE THREAD

**Issue by benlongo** [drwpow/openapi-typescript#1105]
**Description**

Discriminated unions with `oneOf` result in the discriminator being applied to child properties of the variant type instead of as a property of the variant type itself. I'll try to get a minimal repro.

| Name                 | Version                      |
|:---------------------|:-----------------------------|
| `openapi-typescript` | `6.2.4`                      |
| Node.js              | `18.15.0`                      |
| OS + version         | `Windows 11` |

**Reproduction**

Working on a minimal reproduction at the moment.

**Expected result**

The discriminator should be on the base type, not the child properties.

**Checklist**

- [X] I’m willing to open a PR (see [CONTRIBUTING.md](https://github.com/drwpow/openapi-typescript/blob/main/CONTRIBUTING.md))

**Comment by drwpow** [comment]
Thanks for raising. Yes if you’re able to provide a minimal reproducable example I can prioritize a fix for this.

**Comment by benlongo** [comment]
I probably won't have time in the near future to dissect a clean repro from my schema as I had to bail on switching to this from https://github.com/ferdikoomen/openapi-typescript-codegen due to this bug. The schema that generated the errant properties looked very much like the discriminator example from here https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/. From a cursory look through the tests it seems like discriminators aren't tested in combination with oneOf. Seems like there is a decent chance of that example schema getting a repro.

**Comment by drwpow** [comment]
This library does test discriminator with `oneOf` [here](https://github.com/drwpow/openapi-typescript/blob/main/packages/openapi-typescript/test/index.test.ts#L443) and has great support.

Not sure if this applies to your scenario, but when `oneOf` is combined with `anyOf`, the logical union of shapes can be self-contradictory, or too complex to represent. And neither this library nor the schema prohibits you from doing this, IMO it’s better to avoid complex unions for clarity in type generation as well as docs clarity of being able to clearly represent the shape of something.

**Comment by benlongo** [comment]
My mistake - that test does look like it covers this scenario. I didn't have any appearance of `anyOf` in my schema so I don't think that was the cause. Perhaps it happens when you have properties of type object on the variant types? This weekend I can try to modify the test you linked to recreate the issue I was seeing.

**Comment by drwpow** [comment]
Going to close for now, but if you’re able to test on the most recent version, and you can provide a small, reproducable example of the bug, happy to reopen. I believe you were having an issue and probably encountered a bug, but I’m not able to provide a fix without more information.

**Comment by olivierbeaulieu** [comment]
I believe I just ran into this issue, on v6.2.8. Might also exist in v7 - I don't know.

Here's is a failing unit test that reproduces the issue: https://github.com/openapi-ts/openapi-typescript/compare/v6.2.8...olivierbeaulieu:openapi-typescript:nested-discriminators-bug

I don't think `collectCompositions` should be wrapping `BabyPet` in this example with an Omit - it should just inline the schema.

I tried to find the solution, but I haven't found it yet.
